### PR TITLE
Unix domain sockets

### PR DESCRIPTION
**Pull request by ghost**
What is missing:
-I was not able to have config.in working properly
-unix domain socket is not removed if cjet gets killed

### CODE REVIEW

File: src/linux/linux_io.c [hunk 1]
```diff
@@ -110,17 +109,34 @@ static int prepare_peer_socket(int fd)
...
-	if (configure_keepalive(fd) < 0) {
-		log_err("Could not configure keepalive!\n");
-		close(fd);
+	int domain;
+	socklen_t len;
+	if (getsockopt(fd, SOL_SOCKET, SO_DOMAIN, &domain, &len) < 0) {
+		log_err("could not determine socket domain");
+		return -1;
+	}
+	if ((domain==AF_INET) || (domain==AF_INET6)) {
```
Comment (gatzka) [inline src/linux/linux_io.c]: Please adhere to coding style so space around ==

File: src/linux/linux_io.c [hunk 2]
```diff
@@ -335,6 +351,50 @@ static int create_server_socket_all_interfaces(int port)
...
+	struct sockaddr_un serveraddr;
+	memset(&serveraddr, 0, sizeof(serveraddr));
+	serveraddr.sun_family = AF_UNIX;
+	strncpy(serveraddr.sun_path, pPath, sizeof(serveraddr.sun_path)-1);
+	unlink(pPath);
+	if (unlikely(bind(listen_fd, (struct sockaddr *)&serveraddr, sizeof(serveraddr)) < 0)) {
+		log_err("binding unix domain socket failed: '%s!\n", strerror(errno));
+		goto error;
+	}
+	chmod(pPath, 0777); /* everyone should have access */
```
Comment (gatzka) [inline src/linux/linux_io.c]: Is 0666 not sufficient?

File: src/linux/linux_io.c [hunk 3]
```diff
@@ -335,6 +351,50 @@ static int create_server_socket_all_interfaces(int port)
...
+static int create_server_unix_domain_socket(const char *pPath)
+{
+	int listen_fd = socket(AF_UNIX, SOCK_STREAM, 0);
+	if (unlikely(listen_fd < 0)) {
+		log_err("Could not create listen socket!\n");
+		return -1;
+	}
+
+	static const int reuse_on = 1;
+	if (unlikely(setsockopt(listen_fd, SOL_SOCKET, SO_REUSEADDR, &reuse_on,
```
Comment (gatzka) [inline src/linux/linux_io.c:363]: Why do we need SO_REUSEADDR for and UDS?

File: src/linux/linux_io.c [hunk 4]
```diff
@@ -335,6 +351,50 @@ static int create_server_socket_all_interfaces(int port)
...
+
+	if (unlikely(set_fd_non_blocking(listen_fd) < 0)) {
+		log_err("Could not set %s!\n", "O_NONBLOCK");
+		goto error;
+	}
+
+	struct sockaddr_un serveraddr;
+	memset(&serveraddr, 0, sizeof(serveraddr));
+	serveraddr.sun_family = AF_UNIX;
+	strncpy(serveraddr.sun_path, pPath, sizeof(serveraddr.sun_path)-1);
```
Comment (gatzka) [inline src/linux/linux_io.c]: Should be
strncpy(serveraddr.sun_path, pPath, sizeof(serveraddr.sun_path) - 1);

File: src/linux/linux_io.c [hunk 5]
```diff
@@ -625,6 +705,26 @@ static int run_io_all_interfaces(struct eventloop *loop, const struct cmdline_co
...
+			.loop = loop,
+			.sock = uds_fd}};
+
+	ret = start_server(&jet_uds_server.ev);
+	if (ret < 0) {
+		close(uds_fd);
+		return -1;
+	}
+
+
```
Comment (gatzka) [inline src/linux/linux_io.c]: Would be nice to factor out that stuff into a function to avoid code duplication.

File: src/linux/linux_io.c [hunk 6]
```diff
@@ -567,6 +627,26 @@ static int run_io_only_local(struct eventloop *loop, const struct cmdline_config
...
 	ret = run_jet(loop, config);
 
 	stop_server(&ipv4_http_server.ev);
```
Comment (gatzka) [inline src/linux/linux_io.c:670]: I miss a function call like
stop_server(&jet_uds_server.ev)

File: src/linux/linux_io.c [hunk 7]
```diff
@@ -665,6 +765,7 @@ int run_io(struct eventloop *loop, const struct cmdline_config *config)
 	} else {
 		ret = run_io_all_interfaces(loop, config, handler, ARRAY_SIZE(handler));
 	}
+	unlink(UDS_PATH);
```
Comment (gatzka) [inline src/linux/linux_io.c]: I would put the unlink to the end of run_io_only_local() and run_io_all_interfaces().

**Review by gatzka — CHANGES_REQUESTED** [review]
Overall, nice and an improvement.

File: src/linux/linux_io.c (hunk 1 above)
Comment (ghost) [reply src/linux/linux_io.c]: done

File: src/linux/linux_io.c (hunk 2 above)
Comment (ghost) [reply src/linux/linux_io.c]: yes

File: src/linux/linux_io.c (hunk 3 above)
Comment (ghost) [reply src/linux/linux_io.c:363]: If I remove it the second client won't connect...

File: src/linux/linux_io.c (hunk 4 above)
Comment (ghost) [reply src/linux/linux_io.c]: done

File: src/linux/linux_io.c (hunk 5 above)
Comment (ghost) [reply src/linux/linux_io.c]: done

File: src/linux/linux_io.c (hunk 6 above)
Comment (ghost) [reply src/linux/linux_io.c:670]: done

File: src/linux/linux_io.c (hunk 7 above)
Comment (ghost) [reply src/linux/linux_io.c]: done

**Pull request merged**